src/x/instrument/config_prometheus.go

```diff
@@ -0,0 +1,323 @@
+		switch c.OnError {
+		case "stderr":
+			opts.OnRegisterError = func(err error) {
+				fmt.Fprintf(os.Stderr, "tally prometheus reporter error: %v\n", err)
+			}
+		case "log":
+			opts.OnRegisterError = func(err error) {
+				log.Printf("tally prometheus reporter error: %v\n", err)
+			}
+		case "none":
+			opts.OnRegisterError = func(err error) {}
+		default:
+			opts.OnRegisterError = func(err error) {
+				panic(err)
+			}
+		}
+	}
+
+	switch c.TimerType {
+	case "summary":
+		opts.DefaultTimerType = prometheus.SummaryTimerType
+	case "histogram":
+		opts.DefaultTimerType = prometheus.HistogramTimerType
+	}
+
+	if len(c.DefaultHistogramBuckets) > 0 {
+		var values []float64
+		for _, value := range c.DefaultHistogramBuckets {
+			values = append(values, value.Upper)
+		}
+		opts.DefaultHistogramBuckets = values
+	}
+
+	if len(c.DefaultSummaryObjectives) > 0 {
+		values := make(map[float64]float64)
+		for _, value := range c.DefaultSummaryObjectives {
+			values[value.Percentile] = value.AllowedError
+		}
+		opts.DefaultSummaryObjectives = values
+	}
+
+	reporter := prometheus.NewReporter(opts)
+
+	path := "/metrics"
+	if handlerPath := strings.TrimSpace(c.HandlerPath); handlerPath != "" {
+		path = handlerPath
+	}
+
+	handler := reporter.HTTPHandler()
+	if configOpts.Registry != nil {
+		gatherer := newMultiGatherer(configOpts.Registry, configOpts.ExternalRegistries)
+		handler = promhttp.HandlerFor(gatherer, promhttp.HandlerOpts{})
+	}
+
+	if addr := strings.TrimSpace(c.ListenAddress); addr == "" {
+		http.Handle(path, handler)
+	} else {
+		mux := http.NewServeMux()
+		mux.Handle(path, handler)
+		go func() {
+			if err := http.ListenAndServe(addr, mux); err != nil {
+				opts.OnRegisterError(err)
+			}
+		}()
+	}
+
+	return reporter, nil
+}
+
+func newMultiGatherer(
+	primary *prom.Registry,
+	ext []PrometheusExternalRegistry,
+) prom.Gatherer {
+	return &multiGatherer{
+		primary: primary,
+		ext:     ext,
+	}
+}
+
+var _ prom.Gatherer = (*multiGatherer)(nil)
+
+type multiGatherer struct {
+	primary *prom.Registry
+	ext     []PrometheusExternalRegistry
+}
+
+func (g *multiGatherer) Gather() ([]*dto.MetricFamily, error) {
+	results, err := g.primary.Gather()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(g.ext) == 0 {
+		return results, nil
+	}
+
+	for _, secondary := range g.ext {
+		gathered, err := secondary.Registry.Gather()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, elem := range gathered {
+			entry := &dto.MetricFamily{
+				Name:   elem.Name,
+				Help:   elem.Help,
+				Metric: make([]*dto.Metric, 0, len(elem.Metric)),
+			}
+
+			if v := elem.Type; v != nil {
+				metricType := dto.MetricType(*v)
+				entry.Type = &metricType
+			}
+
+			for _, metricElem := range elem.Metric {
+				metricEntry := &dto.Metric{
+					Label:       make([]*dto.LabelPair, 0, len(metricElem.Label)),
+					TimestampMs: metricElem.TimestampMs,
+				}
+
+				if v := metricElem.Gauge; v != nil {
+					metricEntry.Gauge = &dto.Gauge{
```

Comment:
Can different metrics have multiple metric types or should these continue?

Reply:
I'm really just exporting whatever the exporter is already gathering, it's pure pass through here - think of it as a serialization issue (lol).